### PR TITLE
Do some cleanup.

### PR DESCRIPTION
**Pull request by geeknoid**
- Add Helm to the image.

- Create a python context for clarity

- Remove various things we don't need in the final image

- Move things together that belong together, this requires less global reasoning

### CODE REVIEW

File: docker/build-tools/Dockerfile [hunk 1]
```diff
@@ -101,59 +133,32 @@ RUN GO111MODULE=on go get istio.io/tools/protoc-gen-docs
...
+RUN mv /go/bin/* ${OUTDIR}/go/bin
 
 # ShellCheck linter
 ADD https://storage.googleapis.com/shellcheck/shellcheck-${SHELLCHECK_VERSION}.linux.x86_64.tar.xz /tmp
 RUN tar -xJf /tmp/shellcheck-${SHELLCHECK_VERSION}.linux.x86_64.tar.xz -C /tmp
-RUN cp /tmp/shellcheck-${SHELLCHECK_VERSION}/shellcheck ${OUTDIR}/usr/bin/shellcheck
+RUN mv /tmp/shellcheck-${SHELLCHECK_VERSION}/shellcheck ${OUTDIR}/usr/bin
 
-# Hadolint linter
+# Hadolint linte
```
Comment (howardjohn) [inline docker/build-tools/Dockerfile]: typo

File: docker/build-tools/Dockerfile [hunk 2]
```diff
@@ -28,10 +28,10 @@
 # Binary tools
 #############
 
-FROM golang:1.12.5 as binary_tools_context
+FROM golang:1.12.9 as binary_tools_context
 
 # Pinned versions of stuff we pull in
-ENV PROTOBUF_VERSION=3.6.1
```
Comment (ericvn) [inline docker/build-tools/Dockerfile]: Minor nit might be to alphabetize (or similarly group) the versions. Can be done later if desired.

File: docker/build-tools/Dockerfile [hunk 3]
```diff
@@ -240,6 +245,32 @@ RUN gem install --no-wrappers --no-document mdl -v ${MDL_VERSION}
...
+RUN curl "https://bootstrap.pypa.io/get-pip.py" -o "get-pip.py"
+RUN python3 get-pip.py --no-wheel
+
+# Install yamllint
+RUN pip install yamllint==${YAMLLINT_VERSION}
+
+# Clean up stuff we don't need in the final image
+RUN rm -rf /usr/local/bin/pip*
+RUN rm -fr /usr/local/lib/python3.5/dist-packages/pip
+
```
Comment (ericvn) [inline docker/build-tools/Dockerfile]: Another minor nit would be adding the required list for python similar to the others (golang, general, ruby).

File: docker/build-tools/Dockerfile [hunk 4]
```diff
@@ -240,6 +245,32 @@ RUN gem install --no-wrappers --no-document mdl -v ${MDL_VERSION}
 RUN gem install --no-wrappers --no-document html-proofer -v ${HTMLPROOFER_VERSION}
 RUN gem install --no-wrappers --no-document awesome_bot -v ${AWESOMEBOT_VERSION}
 
+#############
```
Comment (ericvn) [inline docker/build-tools/Dockerfile]: Good idea to make its own context.

File: docker/build-tools/Dockerfile [hunk 5]
```diff
@@ -57,30 +59,60 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
     xz-utils
 
 # Install protoc
-WORKDIR /tmp
-ADD https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOBUF_VERSION}/protoc-${PROTOBUF_VERSION}-linux-x86_64.zip /tmp/
-RUN unzip /tmp/protoc-${PROTOBUF_VERSION}-linux-x86_64.zip
+ADD https://github.com/protocolbuffers/protobuf/releases/download/v${PROTOC_VERSION}/protoc-${PROTOC_VERSION}-linux-x86_64.zip /tmp/
+RUN unzip /tmp/protoc-${PROTOC_VERSION}-linux-x86_64.zip
+RUN mkdir -p ${OUTDIR}/usr/bin
+RUN mv /tmp/bin/protoc ${OUTDIR}/usr/bin
```
Comment (sdake) [inline docker/build-tools/Dockerfile]: this changes the install of protoc a little bit - but works for me.

File: docker/build-tools/Dockerfile [hunk 6]
```diff
@@ -57,30 +59,60 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
+RUN git clone --depth 1 --single-branch --branch master https://github.com/kubernetes/apimachinery.git
+WORKDIR /tmp/apimachinery
+RUN find . -name \*proto -exec cp -a --parents {} ${OUTDIR}/usr/include/protobuf/k8s.io/apimachinery \;
+WORKDIR /tmp
+
+RUN mkdir -p ${OUTDIR}/usr/include/protobuf/github.com/gogo/protobuf
+RUN git clone --depth 1 --single-branch --branch master https://github.com/gogo/protobuf.git
+WORKDIR /tmp/protobuf
+RUN find . -name \*proto -exec cp -a --parents {} ${OUTDIR}/usr/include/protobuf/github.com/gogo/protobuf \;
+RUN rm -fr ${OUTDIR}/usr/include/protobuf/github.com/gogo/protobuf/test
```
Comment (sdake) [inline docker/build-tools/Dockerfile]: ina . followup these rm should be separated and comments from the main workflow of the K8s proto clones.

File: docker/build-tools/Dockerfile [hunk 7]
```diff
@@ -57,30 +59,60 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
         done
+
+RUN mkdir -p ${OUTDIR}/usr/include/protobuf/validate
 RUN curl -L -o ${OUTDIR}/usr/include/protobuf/validate/validate.proto https://raw.githubusercontent.com/envoyproxy/protoc-gen-validate/${PROTOC_GEN_VALIDATE_VERSION}/validate/validate.proto
+
+RUN mkdir -p ${OUTDIR}/usr/include/protobuf/gogoproto
 RUN curl -L -o ${OUTDIR}/usr/include/protobuf/gogoproto/gogo.proto https://raw.githubusercontent.com/gogo/protobuf/master/gogoproto/gogo.proto
 
-# Install a bunch of Go tools
+RUN mkdir -p ${OUTDIR}/usr/include/protobuf/k8s.io/api
```
Comment (sdake) [inline docker/build-tools/Dockerfile]: some gold plating would help here explain what this section does in terms of a comment.

File: docker/build-tools/Dockerfile [hunk 8]
```diff
@@ -101,59 +133,32 @@ RUN GO111MODULE=on go get istio.io/tools/protoc-gen-docs
 RUN GO111MODULE=on go get istio.io/tools/cmd/annotations_prep
 RUN GO111MODULE=on go get istio.io/tools/openapi/cue
 
-# Put the stuff we need in its final output location
+# Compress the Go tools and put them in their final location
+ADD https://github.com/upx/upx/releases/download/v${UPX_VERSION}/upx-${UPX_VERSION}-amd64_linux.tar.xz /tmp
+RUN tar -xJf upx-${UPX_VERSION}-amd64_linux.tar.xz -C /tmp
+RUN mv /tmp/upx-${UPX_VERSION}-amd64_linux/upx /usr/bin
+RUN upx --lzma /go/bin/*
```
Comment (sdake) [inline docker/build-tools/Dockerfile]: this isn't compressing all of the tools.  I'd instead compress `${OUTDIR}/usr/bin` (this misses protoc) but can be handled in a followon.

File: docker/build-tools/Dockerfile [hunk 9]
```diff
@@ -240,6 +245,32 @@ RUN gem install --no-wrappers --no-document mdl -v ${MDL_VERSION}
...
+    libc-dev \
+    pkg-config \
+    python3
+
+# Install pip
+RUN curl "https://bootstrap.pypa.io/get-pip.py" -o "get-pip.py"
+RUN python3 get-pip.py --no-wheel
+
+# Install yamllint
+RUN pip install yamllint==${YAMLLINT_VERSION}
```
Comment (sdake) [inline docker/build-tools/Dockerfile]: good change - removing user, but I don't think this should be installed in /usr/local, instead it should go to `/usr/lib/python3.5/site-packages/*` as a followup.

File: docker/build-tools/Dockerfile [hunk 10]
```diff
@@ -302,6 +324,8 @@ COPY --from=ruby_tools_context /usr/local/bin /usr/local/bin
 COPY --from=ruby_tools_context /usr/local/lib /usr/local/lib
 COPY --from=nodejs_tools_context /usr/local /usr/local
 COPY --from=nodejs_tools_context /node_modules/ /node_modules/
+COPY --from=python_context /usr/local/bin /usr/local/bin
```
Comment (sdake) [inline docker/build-tools/Dockerfile]: this should be /usr/lib - copying /usr/local/bin will leave the python3.5 runtime behind...

Comment (sdake) [reply docker/build-tools/Dockerfile]: and /usr/bin :)

**Pull request merged**